mono/mini/mini-llvm.c

```diff
@@ -5023,10 +5047,21 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
 			} else if (ins->opcode == OP_RCOMPARE) {
 				cmp = LLVMBuildFCmp (builder, fpcond_to_llvm_cond [rel], convert (ctx, lhs, LLVMFloatType ()), convert (ctx, rhs, LLVMFloatType ()), "");
 			} else if (ins->opcode == OP_COMPARE_IMM) {
-				if (LLVMGetTypeKind (LLVMTypeOf (lhs)) == LLVMPointerTypeKind && ins->inst_imm == 0)
-					cmp = LLVMBuildICmp (builder, cond_to_llvm_cond [rel], lhs, LLVMConstNull (LLVMTypeOf (lhs)), "");
-				else
-					cmp = LLVMBuildICmp (builder, cond_to_llvm_cond [rel], convert (ctx, lhs, IntPtrType ()), LLVMConstInt (IntPtrType (), ins->inst_imm, FALSE), "");
+				LLVMIntPredicate llvm_pred = cond_to_llvm_cond [rel];
+				if (LLVMGetTypeKind (LLVMTypeOf (lhs)) == LLVMPointerTypeKind && ins->inst_imm == 0) {
+					// We are emitting a NULL check for a pointer
+					gboolean nonnull = mono_llvm_is_nonnull (lhs);
+
+					if (nonnull && llvm_pred == LLVMIntEQ)
+						cmp = LLVMConstInt (LLVMInt1Type (), FALSE, FALSE);
```

Comment:
why do we have to do this ourselves?  does LLVM not evaluate a comparison of a nonnull ptr and 0 to a constant boolean on its own?

Reply:
I observed it not to be spotting that simplification on it's own. When I'm looking for it, I see multiple null checks on wasm being collapsed into a single "if true" branch when I have the above simplification. Otherwise it doesn't happen.

Now what's strange is that if the variable being compared is a function argument with the nonnull attribute on it, LLVM will do the optimization. 

It may be something we could get LLVM to do. An advantage of emitting a simple "true" though is that the bitcode we give LLVM is going to be simpler. When we have explicit null checks, that could amount to a lot of optimizations we aren't forcing LLVM to do before it can accurately see the control flow. In aggregate it probably helps compilation time a little bit too.